lighthouse-core/gather/computed/page-dependency-graph.js

```diff
@@ -34,52 +38,202 @@ class PageDependencyGraphArtifact extends ComputedArtifact {
+      if (
+        evt.name !== TracingProcessor.SCHEDULABLE_TASK_TITLE ||
+        !evt.dur ||
+        evt.dur < MINIMUM_TASK_DURATION
+      ) {
+        i++;
+        continue;
+      }
+
+      // Capture all events that occurred within the task
+      const children = [];
+      i++; // Start examining events after this one
+      for (
+        const endTime = evt.ts + evt.dur;
+        i < traceOfTab.mainThreadEvents.length && traceOfTab.mainThreadEvents[i].ts < endTime;
+        i++
+      ) {
+        children.push(traceOfTab.mainThreadEvents[i]);
+      }
+
+      nodes.push(new CPUNode(evt, children));
+    }
+
+    return nodes;
+  }
+
+  /**
+   * @param {!Node} rootNode
+   * @param {!NetworkNodeOutput} networkNodeOutput
+   */
+  static linkNetworkNodes(rootNode, networkNodeOutput) {
+    networkNodeOutput.nodes.forEach(node => {
+      const initiators = PageDependencyGraphArtifact.getNetworkInitiators(node.record);
       if (initiators.length) {
         initiators.forEach(initiator => {
-          const parent = urlToNodeMap.get(initiator) || rootNode;
+          const parentCandidates = networkNodeOutput.urlToNodeMap.get(initiator) || [rootNode];
+          // Only add the initiator relationship if the initiator is unambiguous
+          const parent = parentCandidates.length === 1 ? parentCandidates[0] : rootNode;
           parent.addDependent(node);
         });
-      } else if (record !== rootRequest) {
+      } else if (node !== rootNode) {
         rootNode.addDependent(node);
       }
     });
+  }
+
+  /**
+   * @param {!Node} rootNode
+   * @param {!NetworkNodeOutput} networkNodeOutput
+   * @param {!Array<!CPUNode>} cpuNodes
+   * @param {!TraceOfTabArtifact} traceOfTab
+   */
+  static linkCPUNodes(rootNode, networkNodeOutput, cpuNodes) {
+    function addDependentNetworkRequest(cpuNode, reqId) {
+      const networkNode = networkNodeOutput.idToNodeMap.get(reqId);
+      if (!networkNode || networkNode.resourceType !== 'xhr') return;
+      networkNode.addDependency(cpuNode);
+    }
+
+    function addDependencyOnUrl(cpuNode, url) {
+      const candidates = networkNodeOutput.urlToNodeMap.get(url) || [];
+
+      let minCandidate = null;
+      let minDistance = Infinity;
+      // Find the closest request that finished before this CPU task started
+      candidates.forEach(candidate => {
+        const distance = cpuNode.startTime - candidate.endTime;
+        if (distance > 0 && distance < minDistance) {
+          minCandidate = candidate;
+          minDistance = distance;
+        }
+      });
+
+      if (!minCandidate) return;
+      cpuNode.addDependency(minCandidate);
+    }
+
+    const timers = new Map();
+    for (const node of cpuNodes) {
+      for (const evt of node.children) {
+        if (!evt.args.data) continue;
+
+        const url = evt.args.data.url;
+        const stackTraceUrls = (evt.args.data.stackTrace || []).map(l => l.url).filter(Boolean);
+
+        switch (evt.name) {
```

Comment:
A few other events that mighttttttttt be worthwhile?  


actually here's a list of main thread trace events with a url in the data:

       1 "name":"CommitLoad"
       8 "name":"EvaluateScript"
      26 "name":"FunctionCall"
       1 "name":"InvalidateLayout"
       2 "name":"Layout"
     179 "name":"PaintImage"
       7 "name":"ParseHTML"
       1 "name":"RenderFrameImpl::didStartProvisionalLoad"
      61 "name":"ResourceSendRequest"
       3 "name":"ScheduleStyleRecalculation"
      12 "name":"TimerInstall"
       7 "name":"TimerRemove"
       1 "name":"TracingStartedInPage"
       2 "name":"UpdateLayoutTree"
       1 "name":"XHRLoad"
       4 "name":"XHRReadyStateChange"
       9 "name":"v8.compile"

    # via: ag '61815.*?775.*?"name":".*?".*?"url":".*?"' news.google.com_2017-09-01_18-14-51-0.trace.json -o  | ag -o '"name":".*?"' | sort | uniq -c


* `ParseHTML` seems good
* `PaintImage` seems good?
* `UpdateLayoutTree` sometimes has a stack trace, which indicates a forced layout. Not sure if the semantics of that matches...
* DecodeImage and ResizeImage i think happen off the main thread..

Reply:
I'm hesitant to add reliance on PaintImage since it seems to have a larger "unthrottled view obfuscates relationships" effect, i.e. the most common reason PaintImage is included in this task is because the image happened to finish downloading then not because this task necessarily *relied* on the image being available.

UpdateLayoutTree, ParseHTML I can add 👍

Reply:
SGTM. as discussed offline, right now the graph is identifying which work is DEFINITELY required before each node.  Images don't really fall into this definition. 
Later work could introduce a pessimistic graph where relationships like these images being painted are included. But i agree it's unclear that would add value.